components/autocomplete/index.js

```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
 		this.search = this.search.bind( this );
 		this.setSelectedIndex = this.setSelectedIndex.bind( this );
 
 		this.state = this.constructor.getInitialState();
 	}
 
-	bindMenuNode( node ) {
-		this.menuNode = node;
+	bindNode( node ) {
+		this.node = node;
+	}
+
+	replace( range, replacement ) {
+		const container = document.createElement( 'div' );
+		container.innerHTML = renderToString( replacement );
+		while ( container.firstChild ) {
+			const child = container.firstChild;
+			container.removeChild( child );
+			range.insertNode( child );
+			range.setStartAfter( child );
+		}
+		range.deleteContents();
 	}
 
 	select( option ) {
-		const { onSelect } = this.props;
+		const { open, range, query } = this.state;
+		const { onSelect } = open || {};
 
 		this.reset();
 
 		if ( onSelect ) {
-			onSelect( option );
+			const replacement = onSelect( option.value, range, query );
+			if ( replacement !== undefined ) {
+				this.replace( range, replacement );
+			}
 		}
 	}
 
 	reset() {
 		this.setState( this.constructor.getInitialState() );
 	}
 
-	onBlur( event ) {
-		// Check if related target is not within, in the case that the user is
-		// selecting an option by button click
-		if ( this.menuNode && ! this.menuNode.contains( event.relatedTarget ) ) {
-			this.reset();
+	// this method is separate so it can be overrided in tests
+	getCursor( container ) {
+		const selection = window.getSelection();
+		if ( selection.isCollapsed && container.contains( selection.anchorNode ) ) {
+			return {
+				node: selection.anchorNode,
+				offset: selection.anchorOffset,
+			};
 		}
+		return null;
 	}
 
-	search( event ) {
-		const { triggerPrefix } = this.props;
-		const { isOpen } = this.state;
-
-		const value = event.target.textContent;
+	// this method is separate so it can be overrided in tests
+	createRange( startNode, startOffset, endNode, endOffset ) {
+		const range = document.createRange();
+		range.setStart( startNode, startOffset );
+		range.setEnd( endNode, endOffset );
+		return range;
+	}
 
-		// Currently search supports only if input's entire value is match
-		let pattern = '^';
+	loadOptions( index ) {
+		this.props.completers[ index ].getOptions().then( ( options ) => {
+			this.setState( {
+				[ 'options_' + index ]: options.map(
+					( option, i ) => ( { ...option, key: index + '_' + i } ) ),
+			} );
+		} );
+	}
 
-		if ( triggerPrefix ) {
-			// Trigger prefix may contain reserved characters of a regular
-			// expression, so escape
-			pattern += escapeStringRegexp( triggerPrefix );
+	findMatch( container, cursor, allCompleters, wasOpen ) {
+		const allowAnything = () => true;
+		let endTextNode;
+		let endIndex;
+		let completers = allCompleters.map( ( completer, idx ) => ( { ...completer, idx } ) );
+		if ( wasOpen ) {
+			// put the open completer at the start so it has priority
+			completers = [
+				wasOpen,
+				...completers.filter( ( completer ) => completer.idx !== wasOpen.idx ),
```

Comment:
Do we need the `Array#map` and `Array#filter` ? Wondering if we could omit by the reference directly, since this is run on every keystroke in editable:

```js
let completers = allCompleters;
if ( wasOpen ) {
	completers = [
		wasOpen,
		...without( completers, wasOpen )
	];
}
```

https://lodash.com/docs/4.17.4#without

Reply:
The 'without' method won't work here because it will use object equality which won't be true.

Reply:
Relating to your later comment about the performance of `Array#filter` I will change this to lodash's filter.